### ISSUE TITLE
[BUG] cast error in curry with partial application

### ISSUE THREAD

**Issue by rickyvetter** [ocaml-wasm/wasm_of_ocaml#38]
**Describe the bug**
I am running into an error that is quite hard to build a minimal repro of. I'm still trying to create a standalone reproduction, but I wanted to share here in case it's obvious which piece(s) come together to create this issue. It's with calling [`create_from_uri_option`](https://github.com/janestreet/async_js/blob/master/src/persistent_connection.mli#L18) with code like:

```ocaml
Async_js.Persistent_connection.Rpc.create_from_uri_option
      ~server_name:"test"
      ~connect:(fun (_ : Uri.t option) -> (* we never get to this point anyway *) assert false)
      (fun () -> Deferred.Or_error.return None)
```

Of note, this code runs just fine (or at least until the assert which is expected):

```ocaml
Async_js.Persistent_connection.Rpc.create
      ~server_name:"test"
      ~address:(module Unit)
      ~connect:(fun () -> assert false)
      (fun () -> Deferred.Or_error.return ())
```

Also of note, code like this succeeds:

```ocaml
let create_from_uri_option = Async_js.Persistent_connection.Rpc.create_from_uri_option in
Firebug.console##log create_from_uri_option;
create_from_uri_option
      ~server_name:"test"
      ~connect:(fun (_ : Uri.t option) -> assert false)
      (fun () -> Deferred.Or_error.return None)
```

This implies to me that it has something to do with partial application, and potentially includes, but I can't seem to replicate this either as a single file by hiding with signatures or across files with mlis. But it happens with this specific function reliably.

The error I get is:

```
inline_tests_runner.bc.wasm.js:90701
           throw: e=>{throw e;},
                      ^

RuntimeError: illegal cast
    at curry_7_k5s_ (wasm://wasm/01f728c2:wasm-function[5878]:0x3c7df1)
    ...
```

**Expected behavior**
I would expect this code to work the same as in JavaScript and for there to not be a cast error.

**Versions**
Version of packages used to reproduce the bug
https://github.com/janestreet/async_js master, wasm_of_ocaml d4c5423b8faa59d371ac0f2ec456920cfbf0a301

**Comment by vouillon** [comment]
This is really weird. The definition of `$curry_7` is something like:
```wasm
(func $curry_7 (param $x (ref eq)) (param $f (ref eq))
  (struct.new $curry_7_7
   (ref.func $curry_7_6)
   (ref.cast (ref $closure_7) (local.get $f))
   (local.get $0)))
```
It gets called like this:
```wasm
(call_ref $function_1
  (local.get $x)
  (local.get $f)
  (struct.get $closure 0 (ref.cast (ref $closure) (local.get $f))))
```
The point is that `$curry_7` is called since the first field of function `$f` points to it. But this only happens when `$f` has type `$closure_7`. So I don't really see how the cast can fail.

Another point is that neither `Async_js` nor `Persistent_connection_kernel` contain a function with 7 parameters. `Async` contains a single function with 7 parameters: `Monitor.try_with`. `Async_js.Persistent_connection.Rpc.create_from_uri_option` is a partial application of this function which has 8 parameters:
https://github.com/janestreet/async_kernel/blob/fae6e563424ab2b9f319e24e0bf54b8bcaf2f9e9/persistent_connection_kernel/src/persistent_connection_kernel.ml#L186

If you can reproduce the issue in Chrome, you can put a breakpoint on the `cast` and inspect the values and the stack trace.

**Comment by rickyvetter** [comment]
Sharing a bit more:

```wasm
;; type2 and type92 aren't defined anywhere in this file. not sure if that's expected?
(type $type111 (sub $type3 (struct (field $field0 (ref $type2)) (field $field1 (ref $type92)))))
;; ...
  (func $curry_7_k5w_  (param $0 (ref eq)) (param $1 (ref eq)) (result (ref eq))
    ref.func $curry_7_6_tSG_
    local.get $1
    ref.cast $type111 ;; error on this line
    local.tee $1
    ref.cast $type111
    local.get $0
    struct.new $type181
  )
```
When I break on the exception, `$1` and `$0` are both:

```js
{
    "type": "(ref eq)",
    "value": 0
}
```

Isn't `create_from_uri_option` a function with 7 paramaters? Since it's already applied `~address`?

The caller of curry_7 is pretty suspect as well:

```wasm
  (func $_vXU_ (param $0 (ref eq)) (param $1 (ref eq)) (result (ref eq))
    (local $2 (ref $type65))
    (local $4 (ref eq))
    (local $7 (ref eq))
    (local $9 (ref eq))
    (local $11 (ref eq))
    (local $14 (ref eq))
    (local $16 (ref eq))
    (local $18 (ref eq))
    (local $20 (ref eq))
    local.get $1
    ref.cast $type65
    local.tee $1
    ref.cast $type65
    local.set $2
    nop
    nop
    nop
    nop
    nop
    global.get $global10445
    global.get $global10446
    i32.const 0
    ref.i31
    i32.const 0
    ref.i31
    ref.func $_vVx_
    i32.const 0
    ref.i31
    i32.const 0
    ref.i31
    i32.const 0
    ref.i31
    call $curry_7_k5w_
    ;; ...
  ```
I originally repro'd with a build of wasm_of_ocaml that passed these flags to binaryen `--simplify-locals-notee-nostructure --reorder-locals`, but it seems I'm also able to repro with the flags in the repo today, so I don't think those matter.

**Comment by vouillon** [comment]
> Isn't `create_from_uri_option` a function with 7 paramaters? Since it's already applied `~address`?

Oh, indeed. Wasm_of_ocaml is certainly doing an eta-expansion. I did not saw that since I was looking at the output of separate compilation.

> The caller of curry_7 is pretty suspect as well:
> 
> ```webassembly
>   (func $_vXU_ (param $0 (ref eq)) (param $1 (ref eq)) (result (ref eq))
>    [...]
>     global.get $global10445
>     global.get $global10446
>     i32.const 0
>     ref.i31
>     i32.const 0
>     ref.i31
>     ref.func $_vVx_
>     i32.const 0
>     ref.i31
>     i32.const 0
>     ref.i31
>     i32.const 0
>     ref.i31
>     call $curry_7_k5w_
>     ;; ...
> ```

So `curry_7_k5w_` is called with completely incorrect arguments. The values `(ref.i31 (const 0))` probably correspond to `None`.

What is the definition of `$_vXU_` in the `.wat` file? The following instructions could be interesting as well. Many values are pushed on the stack. `curry_7_k5w_` is consuming only two of them.

**Comment by rickyvetter** [comment]
Here is the full definition pulled from Chrome:

```wasm
(func $_vXU_  (param $0 (;0;) (ref eq)) (param $1 (ref eq)) (result (ref eq))
  (local $2 (ref $type65))
  (local $4 (ref eq))
  (local $7 (ref eq))
  (local $9 (ref eq))
  (local $11 (ref eq))
  (local $14 (ref eq))
  (local $16 (ref eq))
  (local $18 (ref eq))
  (local $20 (ref eq))
  local.get $1
  ref.cast $type65
  local.tee $1
  ref.cast $type65
  local.set $2
  nop
  nop
  nop
  nop
  nop
  global.get $global10445
  global.get $global10446
  i32.const 0
  ref.i31
  i32.const 0
  ref.i31
  ref.func $_vVx_
  i32.const 0
  ref.i31
  i32.const 0
  ref.i31
  i32.const 0
  ref.i31
  call $curry_7_k5w_
  local.tee $4
  local.get $4
  ref.cast $type3
  local.tee $4
  ref.cast $type3
  struct.get $type3 $field0
  call_ref $type2
  struct.new $type5
  local.tee $7
  local.get $7
  ref.cast $type3
  local.tee $7
  ref.cast $type3
  struct.get $type3 $field0
  call_ref $type2
  local.tee $9
  local.get $9
  ref.cast $type3
  local.tee $9
  ref.cast $type3
  struct.get $type3 $field0
  call_ref $type2
  call $apply_2_tRZ_
  local.set $11
  nop
  nop
  ref.func $_vWO_
  i32.const 50
  ref.i31
  i32.const 0
  ref.i31
  call $of_int_ms_m6v_
  local.get $0
  struct.new $type6
  local.set $14
  nop
  ref.func $_vWP_
  ref.func $_vXk_
  i32.const 1
  ref.i31
  local.get $2
  struct.get $type65 $field15
  array.new_fixed $block 2
  struct.new $type5
  struct.new $type5
  local.set $16
  nop
  ref.func $connector_vXq_
  local.get $2
  struct.get $type65 $field15
  local.get $11
  struct.new $type6
  global.get $global206
  global.get $global206
  ref.cast $type3
  struct.get $type3 $field0
  call_ref $type2
  local.set $18
  nop
  ref.func $_vXr_
  local.get $16
  local.get $14
  struct.new $type6
  local.get $2
  struct.get $type65 $field17
  i32.const 0
  ref.i31
  call $handle_nSI_
  local.set $20
  nop
  nop
  nop
  nop
  global.get $global18125
  i32.const 0
  ref.i31
  i32.const 0
  ref.i31
  local.get $2
  struct.get $type65 $field17
  global.get $global2434
  local.get $18
  local.get $20
  i32.const 0
  ref.i31
  call $store_nRT_
  i32.const 0
  ref.i31
  call $perform_nSA_
  global.get $global10443
  i32.const 0
  ref.i31
  call $map_nTv_
  ref.func $_vXT_
  local.get $2
  struct.get $type65 $field1
  local.get $2
  struct.get $type65 $field2
  local.get $2
  struct.get $type65 $field3
  local.get $2
  struct.get $type65 $field4
  local.get $2
  struct.get $type65 $field5
  local.get $2
  struct.get $type65 $field6
  local.get $2
  struct.get $type65 $field7
  local.get $2
  struct.get $type65 $field8
  local.get $2
  struct.get $type65 $field9
  local.get $2
  struct.get $type65 $field10
  local.get $2
  struct.get $type65 $field11
  local.get $2
  struct.get $type65 $field12
  local.get $2
  struct.get $type65 $field13
  local.get $2
  struct.get $type65 $field14
  local.get $2
  struct.get $type65 $field16
  local.get $2
  struct.get $type65 $field17
  struct.new $type48
  i32.const 0
  ref.i31
  return_call $sub_nTB_
)
```

Here's the function in the `.wat` file:

```wasm
(func
  $_vXU_
  (param (ref eq) (ref eq))
  (result (ref eq))
  (local (ref $env_1_17_rdQ_) (ref eq) (ref eq) (ref eq) (ref eq) (ref eq)
    (ref eq) (ref eq) (ref eq) (ref eq) (ref eq) (ref eq) (ref eq) 
    (ref eq) (ref eq) (ref eq) (ref eq) (ref eq) (ref eq) (ref eq) 
    (ref eq) (ref eq) (ref eq) (ref eq))
  (local.set 2
    (ref.cast (ref $env_1_17_rdQ_)
      (local.tee 1 (ref.cast (ref $env_1_17_rdQ_) (local.get 1)))))
  (local.set 3
    (call $curry_7_k5w_ (ref.i31 (i32.const 0)) (ref.i31 (i32.const 0))))
  (local.set 5
    (call_ref $function_1_kYB_ (ref.i31 (i32.const 0))
      (local.tee 4 (local.get 3))
      (struct.get $closure_kYC_ 0
        (ref.cast (ref $closure_kYC_)
          (local.tee 4 (ref.cast (ref $closure_kYC_) (local.get 4)))))))
  (local.set 6 (struct.new $env_1_1_k7W_ (ref.func $_vVx_) (local.get 5)))
  (local.set 8
    (call_ref $function_1_kYB_ (ref.i31 (i32.const 0))
      (local.tee 7 (local.get 6))
      (struct.get $closure_kYC_ 0
        (ref.cast (ref $closure_kYC_)
          (local.tee 7 (ref.cast (ref $closure_kYC_) (local.get 7)))))))
  (local.set 10
    (call_ref $function_1_kYB_ (ref.i31 (i32.const 0))
      (local.tee 9 (local.get 8))
      (struct.get $closure_kYC_ 0
        (ref.cast (ref $closure_kYC_)
          (local.tee 9 (ref.cast (ref $closure_kYC_) (local.get 9)))))))
  (local.set 11
    (call $apply_2_tRZ_ (global.get $_rd0_) (global.get $_rdY_)
      (local.get 10)))
  (local.set 12
    (call $of_int_ms_m6v_ (ref.i31 (i32.const 50)) (ref.i31 (i32.const 0))))
  (local.set 13
    (array.new_fixed $block_a_ 2 (ref.i31 (i32.const 1))
      (struct.get $env_1_17_rdQ_ 15 (local.get 2))))
  (local.set 14
    (struct.new $env_1_2_las_ (ref.func $_vWO_) (local.get 12)
      (local.get 0)))
  (local.set 15
    (struct.new $env_1_1_k7W_ (ref.func $_vXk_) (local.get 13)))
  (local.set 16
    (struct.new $env_1_1_k7W_ (ref.func $_vWP_) (local.get 15)))
  (local.set 17
    (struct.new $env_1_2_las_ (ref.func $connector_vXq_)
      (struct.get $env_1_17_rdQ_ 15 (local.get 2)) (local.get 11)))
  (local.set 18
    (call_ref $function_1_kYB_ (local.get 17) (global.get $return_nTT_)
      (struct.get $closure_kYC_ 0
        (ref.cast (ref $closure_kYC_) (global.get $return_nTT_)))))
  (local.set 19
    (struct.new $env_1_2_las_ (ref.func $_vXr_) (local.get 16)
      (local.get 14)))
  (local.set 20
    (call $handle_nSI_ (local.get 19)
      (struct.get $env_1_17_rdQ_ 17 (local.get 2)) (ref.i31 (i32.const 0))))
  (local.set 21
    (call $store_nRT_ (global.get $connector_var_n1D_) (local.get 18)
      (local.get 20) (ref.i31 (i32.const 0))))
  (local.set 22
    (call $perform_nSA_ (ref.i31 (i32.const 0))
      (struct.get $env_1_17_rdQ_ 17 (local.get 2)) (local.get 21)
      (ref.i31 (i32.const 0))))
  (local.set 23
    (call $map_nTv_ (ref.i31 (i32.const 0)) (local.get 22)
      (global.get $_rd6_) (ref.i31 (i32.const 0))))
  (local.set 24
    (struct.new $env_1_16_rd7_ (ref.func $_vXT_)
      (struct.get $env_1_17_rdQ_ 1 (local.get 2))
      (struct.get $env_1_17_rdQ_ 2 (local.get 2))
      (struct.get $env_1_17_rdQ_ 3 (local.get 2))
      (struct.get $env_1_17_rdQ_ 4 (local.get 2))
      (struct.get $env_1_17_rdQ_ 5 (local.get 2))
      (struct.get $env_1_17_rdQ_ 6 (local.get 2))
      (struct.get $env_1_17_rdQ_ 7 (local.get 2))
      (struct.get $env_1_17_rdQ_ 8 (local.get 2))
      (struct.get $env_1_17_rdQ_ 9 (local.get 2))
      (struct.get $env_1_17_rdQ_ 10 (local.get 2))
      (struct.get $env_1_17_rdQ_ 11 (local.get 2))
      (struct.get $env_1_17_rdQ_ 12 (local.get 2))
      (struct.get $env_1_17_rdQ_ 13 (local.get 2))
      (struct.get $env_1_17_rdQ_ 14 (local.get 2))
      (struct.get $env_1_17_rdQ_ 16 (local.get 2))
      (struct.get $env_1_17_rdQ_ 17 (local.get 2))))
  (return_call $sub_nTB_ (global.get $const_kzf_) (local.get 23)
    (local.get 24) (ref.i31 (i32.const 0))))
  ```

**Comment by vouillon** [comment]
Thanks a lot! That should be fixed by #40.
